Run akka-remote tests on Scala 3

raboof opened the pull request:
refs #30243

raboof committed:
Run akka-remote tests on Scala 3

octonato commented on .travis.yml:
```diff
@@ -38,7 +38,7 @@ jobs:
       name: scala3
       # separate job since only a few modules compile with Scala 3 yet
-      script: jabba install adopt@1.11-0 && jabba use adopt@1.11-0 && sbt -Dakka.build.scalaVersion=3.0 "akka-actor-tests/test:compile" akka-actor-testkit-typed/compile akka-actor-typed/compile akka-discovery/test akka-pki/test:compile akka-protobuf/test:compile akka-protobuf-v3/test:compile akka-slf4j/test:compile akka-stream/compile akka-stream-tests-tck/test akka-coordination/test akka-seriali ...
+      script: jabba install adopt@1.11-0 && jabba use adopt@1.11-0 && sbt -Dakka.build.scalaVersion=3.0 "akka-actor-tests/test:compile" akka-actor-testkit-typed/compile akka-actor-typed/compile akka-discovery/test akka-pki/test:compile akka-protobuf/test:compile akka-protobuf-v3/test:compile akka-slf4j/test:compile akka-stream/compile akka-stream-tests-tck/test akka-coordination/test akka-seriali ...
```
Now I see that some modules we are running the tests while others we are only `test:compile`. 

Shall we run the tests for them all?

raboof replied:
if they succeed, then yes :)

octonato replied:
Do we have tests failures when compiling with Scala 3?

raboof replied:
No, otherwise travis would fail this job

octonato replied:
I mean the other modules that are still using `test:compile`. I was wondering if they are so because there are some failures which would surprise me, but we never know. 🤷 

Anyway, I will PR and remove all the `test:compile`.

octonato replied:
Well, I just read your comment about the object / class and serialization. Yeah, indeed a good example of compilation  causing runtime difference. Programming is not easy. 😞

octonato commented on akka-remote/src/test/scala/akka/remote/artery/SerializationErrorSpec.scala:
```diff
@@ -41,7 +41,7 @@ class SerializationErrorSpec extends ArteryMultiNodeSpec(ArterySpecSupport.defau
       expectMsg("ping")
 
-      EventFilter[java.io.NotSerializableException](start = "Failed to serialize message", occurrences = 1).intercept {
+      EventFilter[Exception](start = "Failed to serialize message", occurrences = 1).intercept {
```
why not keep `NotSerializableException`?

raboof replied:
Because we got a different exception

ignasi35 replied:
> we got a different exception

regression?

Should the test then be explicit about that other exception?

raboof replied:
Thanks for keeping me honest ;)

This test was still throwing `java.io.NotSerializableException` on Scala 2.13, but not on Scala 3.

This turned out to be because the message is an `object`, which is `java.io.Serializable` on Scala 3, but not on Scala 2.13, leading to a different failure mode. Changing it to a `class`, which is not `java.io.Serializable` on either Scala version, makes it behave the same on both versions again.

octonato reviewed (approved):
LGTM, with some questions

raboof committed:
Test the same scenario on Scala 2.13 and 3

A case object is `Serializable` on Scala 3 but not on Scala 2.13, so
this test would fail because no serializer is found on Scala 2.13,
and because the Java serializer is disabled on Scala 3.

By using a class instead, which is not `Serializable` both on Scala 2.13
and Scala 3, we test the same thing on both Scala versions.

ignasi35 commented on akka-multi-node-testkit/src/main/scala/akka/remote/testkit/MultiNodeSpec.scala:
```diff
@@ -287,7 +285,7 @@ abstract class MultiNodeSpec(
     })
 
-  val log: LoggingAdapter = Logging(system, this.getClass)
+  val log: LoggingAdapter = Logging(system, this)(_.getClass.getName)
```
Hmmm, is this so the logger uses the subclass name as the logger? Nice trick.

ignasi35 commented on akka-remote/src/test/scala/akka/remote/artery/SerializationErrorSpec.scala:
```diff
@@ -11,7 +11,7 @@ import akka.testkit.TestActors
 object SerializationErrorSpec {
 
-  object NotSerializableMsg
+  class NotSerializableMsg
```
hehehe

ignasi35 commented on akka-remote/src/test/scala/akka/remote/classic/transport/AkkaProtocolSpec.scala:
```diff
@@ -395,8 +389,6 @@ class AkkaProtocolSpec extends AkkaSpec("""akka.actor.provider = remote """) wit
           h.localAddress should ===(localAkkaAddress)
           h
-
-        case _ => fail()
```
What's going on here? The `fail()` is no longer necessary because the `MatchError` will fail the test anyway?

(I'm curious) How did the `case _ => fail()` cause scala 3 to complain?

octonato replied:
`val statusPromise: Promise[AssociationHandle]`

So it can only be a `AssociationHandle`. Scala 3 probably complained because `case _` will never happen. 

Which bring us to the question, why are we pattern matching? 😄

raboof replied:
> Scala 3 probably complained because `case _` will never happen.

Yes

> why are we pattern matching?

Hum, I agree it looks like it'd be good to remove the pattern-match entirely... not sure I want to grow this PR with that though :)

ignasi35 reviewed (approved):
LGTM. Added a couple of non-blocking comments.

raboof commented:
(failure is #30424, re-running)

jrudolph reviewed (approved):
LGTM
